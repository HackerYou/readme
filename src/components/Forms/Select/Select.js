import React from 'react';
import PropTypes from 'prop-types';
import keygen from 'keygenerator';
import classNames from 'classnames';

const Select = (props) => {
    const {
        name,
        labelText,
        options,
        chosenKey,
        chosenVal,
        labelInline,
        handleChange,
        classes,
        chosenText,
        value,
    } = props;
    const labelClass = classNames({
        inline: labelInline,
    });
    return (<span>
        {labelText !== '' && <label htmlFor={name} className={labelClass}>{labelText}</label>}
        <select name={name} className={classes} value={value} onChange={handleChange}>
            <option value="default" disabled>Select an option...</option>
            { options.map((option) => {
                if (typeof option === 'string') {
                    return <option key={keygen._()} value={option}>{option}</option>;
                }
                return (
                    <option key={option[chosenKey]} value={option[chosenVal]}>
                        {option[chosenText]}
                    </option>
                );
            })}
        </select>
    </span>);
};

Select.defaultProps = {
    classes: '',
    labelText: '',
    handleChange: null,
    chosenKey: '',
    chosenText: '',
    chosenVal: '',
    labelInline: false,
};

Select.propTypes = {
    value: PropTypes.string.isRequired,
    labelInline: PropTypes.bool,
    classes: PropTypes.string,
    chosenKey: PropTypes.string,
    chosenText: PropTypes.string,
    chosenVal: PropTypes.string,
    name: PropTypes.string.isRequired,
    options: PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
    ).isRequired,
    labelText: PropTypes.string,
    handleChange: PropTypes.func,
};

export default Select;
