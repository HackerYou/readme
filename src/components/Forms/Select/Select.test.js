import React from 'react';
import Select from './Select';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

test('It renders the select component when you pass in array into options prop', () => {
    const props = {
        name: 'Test',
        options: ['blue', 'green', 'darkcyan']
    };
    const component = renderer.create(<Select name={props.name} val={props.val} options={props.options} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('It renders the select component when you pass and object into options prop', () => {
    const props = {
        name: 'Tiff',
        chosenKey: '_id',
        chosenVal: 'breed',
        options: [{petType: 'dog', breed:'bulldog', _id: '12312312', petName: 'fluffy'}, {petType: 'cat', breed:'calico', _id: '123213', petName: 'meow'}]
    };
    const component = renderer.create(<Select name={props.name} chosenKey={props.chosenKey} chosenVal={props.chosenVal} options={props.options}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('It calls the function that gets passed into handle change', () => {
    const props = {
        name: 'Tiff',
        chosenKey: 'petName',
        options: [{petType: 'dog', petName: 'fluffy'}, {petType: 'cat', petName: 'meow'}]
    };

    const mockHandleChange = jest.fn();
    const select = shallow(<Select name={props.name} chosenKey={props.chosenKey} options={props.options} handleChange={mockHandleChange} />);
    select.find('select').simulate('change', {target: {value: 'fluffy'}});
    expect(mockHandleChange).toHaveBeenCalled();
});







