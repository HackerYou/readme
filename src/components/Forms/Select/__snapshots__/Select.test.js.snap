// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It renders the select component when you pass and object into options prop 1`] = `
<span>
  <select
    className=""
    name="Tiff"
    onChange={null}
  >
    <option
      value="bulldog"
    >
      fluffy
    </option>
    <option
      value="calico"
    >
      meow
    </option>
  </select>
</span>
`;

exports[`It renders the select component when you pass in array into options prop 1`] = `
<span>
  <select
    className=""
    name="Test"
    onChange={null}
  >
    <option
      value="blue"
    >
      blue
    </option>
    <option
      value="green"
    >
      green
    </option>
    <option
      value="darkcyan"
    >
      darkcyan
    </option>
  </select>
</span>
`;
